### PR TITLE
Wires Multiport Support

### PR DESCRIPTION
**Pull request by MMaiero**

### CODE REVIEW

**Comment by MMaiero** [comment]
Still a work in progress

**Comment by MMaiero** [comment]
Awaiting review.

The result is something like this:
screen shot 2017-12-19 at 16 20 53

**Comment by pierrepitiot** [comment]
@MMaiero using the diagram attached it works for publishing but the "else" branch seems not working, I never have Logger messages in my log.
capture du 2017-12-20 14-07-44

File: kura/org.eclipse.kura.wire.component.conditional.provider/src/main/java/org/eclipse/kura/internal/wire/conditional/ConditionalScriptInterface.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    private List<WireRecord> inputRecords;
+
+    public ConditionalScriptInterface(WireEnvelope inputWireEnvelope) {
+        if (inputWireEnvelope != null) {
+            this.inputRecords = inputWireEnvelope.getRecords();
+        }
+    }
+
+    public Object getInputRecord(int i, String name) {
+        if (i < 0 || i > this.inputRecords.size()) {
```
Comment (markoer) [inline kura/org.eclipse.kura.wire.component.conditional.provider/src/main/java/org/eclipse/kura/internal/wire/conditional/ConditionalScriptInterface.java]: Probably `i >= this.inputRecords.size()`

**Comment by MMaiero** [comment]
To understand the context: the asset always creates a message with that metric?

**Comment by pierrepitiot** [comment]
also, messages are sent twice

**Review by markoer — COMMENTED** [review]
Only one observation during code review.

**Comment by pierrepitiot** [comment]
no, this is only present when a lora frame has been received from the mote LoRaMote.
Here is the messages displayed by a "standard" logger (not behind conditional):

when nothing has been received:
2017-12-20 13:48:36,889 [DefaultQuartzScheduler_Worker-2] ERROR o.e.k.i.w.a.WireAsset - No data available
2017-12-20 13:48:36,904 [DefaultQuartzScheduler_Worker-2] INFO  o.e.k.i.w.l.Logger - Received WireEnvelope from org.eclipse.kura.wire.WireAsset-1513777694806-76
2017-12-20 13:48:36,909 [DefaultQuartzScheduler_Worker-2] INFO  o.e.k.i.w.l.Logger - Record List content: 
2017-12-20 13:48:36,911 [DefaultQuartzScheduler_Worker-2] INFO  o.e.k.i.w.l.Logger -   Record content: 
2017-12-20 13:48:36,912 [DefaultQuartzScheduler_Worker-2] INFO  o.e.k.i.w.l.Logger -     assetName : Motes

when a frame has been received:
2017-12-20 13:48:56,942 [DefaultQuartzScheduler_Worker-4] INFO  o.e.k.i.w.l.Logger - Received WireEnvelope from org.eclipse.kura.wire.WireAsset-1513777694806-76
2017-12-20 13:48:56,946 [DefaultQuartzScheduler_Worker-4] INFO  o.e.k.i.w.l.Logger - Record List content: 
2017-12-20 13:48:56,946 [DefaultQuartzScheduler_Worker-4] INFO  o.e.k.i.w.l.Logger -   Record content: 
2017-12-20 13:48:56,947 [DefaultQuartzScheduler_Worker-4] INFO  o.e.k.i.w.l.Logger -     LoRaMote : 0027930960FCDB2D000000000000FFFF
2017-12-20 13:48:56,948 [DefaultQuartzScheduler_Worker-4] INFO  o.e.k.i.w.l.Logger -     assetName : Motes
2017-12-20 13:48:56,949 [DefaultQuartzScheduler_Worker-4] INFO  o.e.k.i.w.l.Logger -     LoRaMote_timestamp : 1513777736889

**Comment by MMaiero** [comment]
ok, I'll give it a look to see what's wrong.
When I've done my tests I've not noticed duplicate messages, I'll look also at that

**Comment by markoer** [comment]
To me it looks like `then` always executes while `else` only works when condition is false. Visualization is OK.
image

5s timer:
```
then/true:
2017-12-20 15:21:23,645 [DefaultQuartzScheduler_Worker-9] INFO  o.e.k.i.w.l.Logger - Received WireEnvelope from org.eclipse.kura.wire.Conditional-1513778091228-21
2017-12-20 15:21:23,662 [DefaultQuartzScheduler_Worker-9] INFO  o.e.k.i.w.l.Logger - Record List content: 
2017-12-20 15:21:23,662 [DefaultQuartzScheduler_Worker-9] INFO  o.e.k.i.w.l.Logger -   Record content: 
2017-12-20 15:21:23,663 [DefaultQuartzScheduler_Worker-9] INFO  o.e.k.i.w.l.Logger -     TIMER : 1513779683642
2017-12-20 15:21:23,663 [DefaultQuartzScheduler_Worker-9] INFO  o.e.k.i.w.l.Logger - 
2017-12-20 15:21:28,644 [DefaultQuartzScheduler_Worker-10] INFO  o.e.k.i.w.l.Logger - Received WireEnvelope from org.eclipse.kura.wire.Conditional-1513778091228-21
2017-12-20 15:21:28,652 [DefaultQuartzScheduler_Worker-10] INFO  o.e.k.i.w.l.Logger - Record List content: 
2017-12-20 15:21:28,652 [DefaultQuartzScheduler_Worker-10] INFO  o.e.k.i.w.l.Logger -   Record content: 
2017-12-20 15:21:28,653 [DefaultQuartzScheduler_Worker-10] INFO  o.e.k.i.w.l.Logger -     TIMER : 1513779688642
2017-12-20 15:21:28,653 [DefaultQuartzScheduler_Worker-10] INFO  o.e.k.i.w.l.Logger - 

else/false:
2017-12-20 15:21:58,190 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger - Received WireEnvelope from org.eclipse.kura.wire.Conditional-1513778091228-21
2017-12-20 15:21:58,205 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger - Record List content: 
2017-12-20 15:21:58,206 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger -   Record content: 
2017-12-20 15:21:58,207 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger -     TIMER : 1513779718188
2017-12-20 15:21:58,208 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger - 
2017-12-20 15:21:58,209 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger - Received WireEnvelope from org.eclipse.kura.wire.Conditional-1513778091228-21
2017-12-20 15:21:58,210 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger - Record List content: 
2017-12-20 15:21:58,210 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger -   Record content: 
2017-12-20 15:21:58,211 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger -     TIMER : 1513779718188
2017-12-20 15:21:58,211 [DefaultQuartzScheduler_Worker-6] INFO  o.e.k.i.w.l.Logger - 
```

**Comment by markoer** [comment]
I like the `WARN  o.e.k.i.w.c.Conditional - Exception while performing decision.` in case of a syntax error in condition. I guess something like this would come in handy also with typos in metric names in which case the envelope just disappears.

**Comment by markoer** [comment]
Port name display could be optional/togglable. Otherwise I wouldn't it even display it for most components (it's OK with conditions where you can see which output is which, but only clutters the graph in other cases).

**Comment by MMaiero** [comment]
> Port name display could be optional/togglable. Otherwise I wouldn't it even display it for most components (it's OK with conditions where you can see which output is which, but only clutters the graph in other cases).

Agreed. We have to see if this is possible.

**Comment by pierrepitiot** [comment]
I tested the last fix, messages are now published only once but I still don't have any output on the "else" branch on the logger.

**Comment by MMaiero** [comment]
Could it be that the condition specified is not working in some way? Trying setting false as condition does always trigger the else arc?

**Comment by pierrepitiot** [comment]
no, false is working, it triggers the logger

**Comment by markoer** [comment]
> To me it looks like then always executes while else only works when condition is false. Visualization is OK.

I now believe this assessment was not entirely correct. When I set one logger to quiet and other to verbose, it's apparent that only one of them prints output. But the then branch does it twice. I didn't manage to check if Nicola's update fixed it, yet.

**Comment by nicolatimeus** [comment]
@markoer you are right. Due to a bug the conditional component ends up with two wires connected to a port and one connected to the other. The actual port with duplicate wires is random, it can be either the `then` or the `else` port, it depends on the order in which the downstream components are tracked by the WireAdmin. This should be fixed by last commit.

@pierrepitiot I wasn't able to reproduce, during my tests the conditional dispatches the envelopes correctly. Do you have some instructions to reliably reproduce the issue?

**Comment by pierrepitiot** [comment]
@nicolatimeus actually when I test on a modbus device for example (the EDCK4000) with a threshold condition (like qc>10) it works except that the >10 condition is actually treated as >= because the then branch is true when my counter =10...
The condition that is not working is testing the presence of a variable:
wire.getInputRecord(0,"LoRaMote") != null;

Maybe my condition is not correct? But it works on the "then" branch, meaning that I publish a message when I have a LoRaMote variable in the record.

**Comment by nicolatimeus** [comment]
I was able to reproduce one of the issues, a NPE was thrown by the conditional component if you try to access a non-existing metric.
This applies to the `wire.getInputRecord(0,"LoRaMote") != null;` expression if `LoRaMote` does not exist, in this case during my tests the conditional did not emit any envelop on any port.
This should be fixed now. 

Threshold conditions worked fine during my tests.

**Comment by pierrepitiot** [comment]
Yes, that looks good now! Evrything is working :+1:

**Comment by markoer** [comment]
I tested then/else on the Friday version and it was OK.

File: kura/features/org.eclipse.kura.wire.component.merge/pom.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (markoer) [inline kura/features/org.eclipse.kura.wire.component.merge/pom.xml]:
````
%s/\t/    /g
````

File: kura/features/org.eclipse.kura.wire.component.merge/feature.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (markoer) [inline kura/features/org.eclipse.kura.wire.component.merge/feature.xml]: Copyright header missing.
You tend to use different copyright headers for different file types...

File: kura/features/org.eclipse.kura.wire.component.merge/pom.xml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (markoer) [inline kura/features/org.eclipse.kura.wire.component.merge/pom.xml]: Copyright header missing.

File: kura/org.eclipse.kura.wire.component.merge.provider/pom.xml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+<!--
+    Copyright (c) 2017 Eurotech and/or its affiliates and others
+
+     All rights reserved. This program and the accompanying materials
+     are made available under the terms of the Eclipse Public License v1.0
+     which accompanies this distribution, and is available at
+     http://www.eclipse.org/legal/epl-v10.html
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (markoer) [inline kura/org.eclipse.kura.wire.component.merge.provider/pom.xml]: Tabs to spaces.

**Comment by pierantoniomerlino** [comment]
I'm testing the merge (join) component. I saw that if the wire envelopes contain a metric with the same name, the metric received by the second port wins and it is propagated to the output. 

We should write this somewhere. Maybe in the component ui?

**Review by pierantoniomerlino — APPROVED** [review]

**Pull request merged**